src/frontend/src/binder/bind_context.rs

```diff
@@ -70,29 +72,189 @@ pub struct BindContext {
+            // the group
+            if let Some(group_id) = self.column_group_context.mapping.get(&columns[0]) {
+                let group = self.column_group_context.groups.get(group_id).unwrap();
+                if columns.iter().all(|idx| group.indices.contains(idx)) {
+                    return Ok(Some(*group_id));
+                }
+            }
+        }
+        Ok(None)
+    }
+
+    fn get_indices_with_group_id(&self, group_id: u32, column_name: &String) -> Result<Vec<usize>> {
+        let group = self.column_group_context.groups.get(&group_id).unwrap();
+        if let Some(name) = &group.column_name {
+            debug_assert_eq!(name, column_name);
+        }
+        if let Some(min_nullable) = &group.min_nullable_column {
+            Ok(vec![*min_nullable])
+        } else {
+            // These will be converted to a `COALESCE(col1, col2, ..., coln)`
+            let mut indices: Vec<_> = group.indices.iter().copied().collect();
+            indices.sort(); // ensure a deterministic result
+            Ok(indices)
         }
     }
 
-    pub fn get_index(&self, column_name: &String) -> Result<usize> {
+    pub fn get_unqualified_index(&self, column_name: &String) -> Result<Vec<usize>> {
         let columns = self
             .indexs_of
             .get(column_name)
-            .ok_or_else(|| ErrorCode::ItemNotFound(format!("Invalid column: {}", column_name)))?;
+            .ok_or_else(|| ErrorCode::ItemNotFound(format!("Invalid column: {column_name}")))?;
         if columns.len() > 1 {
+            // If there is some group containing the columns and the ambiguous columns are all in
+            // the group
+            if let Some(group_id) = self.column_group_context.mapping.get(&columns[0]) {
+                let group = self.column_group_context.groups.get(group_id).unwrap();
+                if columns.iter().all(|idx| group.indices.contains(idx)) {
+                    if let Some(min_nullable) = &group.min_nullable_column {
+                        return Ok(vec![*min_nullable]);
+                    } else {
+                        // These will be converted to a `COALESCE(col1, col2, ..., coln)`
+                        return Ok(columns.to_vec());
+                    }
+                }
+            }
             Err(ErrorCode::InternalError(format!("Ambiguous column name: {}", column_name)).into())
         } else {
-            Ok(columns[0])
+            Ok(columns.to_vec())
+        }
+    }
+
+    /// Identifies two columns as being in the same group. Additionally, possibly provides one of
+    /// the columns as being `min_nullable`
+    pub fn add_natural_columns(
+        &mut self,
+        left: usize,
+        right: usize,
+        min_nullable_column: Option<usize>,
+    ) {
+        let l_ref = self.column_group_context.mapping.get(&left).is_some();
+        let r_ref = self.column_group_context.mapping.get(&right).is_some();
+        match (l_ref, r_ref) {
```

Comment:
can match option instead of bool

Reply:
Oh yes, initially I didn't not use group_ids so the value in the hashmap was not cloneable, but here it can be cheaply cloned.